episodes/anonymization.md

```diff
@@ -32,19 +33,352 @@ Metadata elements in imaging, such as patient names and addresses, are often cle
+
+```python
+
+import numpy as np
+import matplotlib.pyplot as plt
+from skimage import io, filters
+
+image = 'data/anonym/identifiable_us.jpg'
+io.imshow(image)
+io.show()
+```
+
+```output
+```
+
+![Image from flikr website published with a permissive lisence.](fig/identifiable_us.jpg){alt='Identifiable ultrasound'}
+
+Write code for two approaches that de-identify (remove the annotations from) the ultrasound image data. 
+
+
+::::::::::::::: solution
+
+## Solution
+
+The image  has identifying metdata. You should identify not only the name, but the date and place as problematic.  You can take three different approaches. One would be to mask the data, the other would be to blur it. First we will show a blurred image:
+
+```python
+
+from skimage.filters import gaussian
+from skimage.color import rgb2gray
+
+image_base = io.imread(image)
+image_base = rgb2gray(image_base)
+sub_im = image_base[0:78,:].copy()
+blur_sub_im = gaussian(sub_im, sigma=9)
+final_image = np.zeros(image_base.shape)
+final_image[0:78,:] = blur_sub_im
+final_image[79:,:]= image_base[79:, :]
+io.imshow(final_image)
+
+```
+
+
+```output
+```
+
+![Image after blurring in one area.](fig/blurred_us.png){alt='Non-Identifiable blurred ultrasound'}
+We could have also just make a simple zero-mask: 
+
+```python
+image_masked = io.imread(image)
+image_masked[0:78,:] = 0
+io.imshow(image_masked)
+
+```
+
+
+```output
+```
+
+![Image after masking in one area.](fig/masked_us.png){alt='Non-Identifiable masked ultrasound'}
+
+Note there are other valid solutions, but these two are very common and straightforward. 
+
+Finally, you could always just chop off the offending part so to speak, and resize:
+
+```python
+from skimage.transform import resize
+
+final_image= image_base[79:, :]
+final_image = resize(final_image,image_base.shape)
+io.imshow(final_image)
+```
+
+```output
+```
+
+![Image after crop and resize.](fig/resize_us1.png){alt='Non-Identifiable cropped ultrasound'}
+:::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::::::::::::::::
+
+You have now seen three approaches to removing some of the visual data on a 2-D image. The same approaches can be taken on a 3D image. 
+
+Let's take another look at an image we used earlier:
+
+```python
+import matplotlib.pyplot as plt
+import SimpleITK as sitk
+from skimage import io
+from ipywidgets import interact, fixed
+from IPython.display import clear_output
+import os
+
+
+sag_image =  sitk.ReadImage("data/sitk/A1_grayT1.nrrd", sitk.sitkFloat32)
+cor_image = sitk.PermuteAxes(sag_image, [2, 1, 0])
+
+# General display function for any two 3D images
+def display_images(image1_z, image2_z, image1_npa, image2_npa, title1="Image 1", title2="Image 2"):
```

Comment:
It would be more intuitive to me to have (image1_z, image1_npa, title1, image2_z, image2_npa, title2) rather than interleaving them, but it's a matter of taste.